File: algorithms/code/leetcode/lc020_valid_parentheses/LC020ValidParentheses.Tests.cs
using Xunit;

namespace AlgorithmDesign.code.leetcode.lc020_valid_parentheses
{
    public class SolutionTests
    {
        [Fact]
        public void test_RE()
        {
            Assert.False(new Solution().IsValid(")"));
        }

        [Fact]
        public void test_RB_RE()
        {
            Assert.True(new Solution().IsValid("()"));
        }

        [Fact]
        public void test_RB_RE_SB_SE_CB_CE()
        {
            Assert.True(new Solution().IsValid("()[]{}"));
        }

        [Fact]
        public void test_RB_SB_RE_SE()
        {
            Assert.False(new Solution().IsValid("([)]"));
        }

        [Fact]
        public void test_RB_CB_RE_CE()
        {
            Assert.False(new Solution().IsValid("({)}"));
        }

        [Fact]
        public void test_SB_RB_SE_RE()
        {
            Assert.False(new Solution().IsValid("[(])"));
        }

        [Fact]
        public void test_RE__SE__CE()
        {
            Assert.False(new Solution().IsValid(")"));
            Assert.False(new Solution().IsValid("]"));
            Assert.False(new Solution().IsValid("}"));
        }

        [Fact]
        public void test_nothing()
        {
            Assert.False(new Solution().IsValid(null));
            Assert.True(new Solution().IsValid(""));
        }
    }
}

File: algorithms/code/leetcode/lc020_valid_parentheses/LC020ValidParenthesesTests.java
package leetcode.lc020_valid_parentheses;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

public final class LC020ValidParenthesesTests {
    @Test
    public void testGarbage() throws Exception {
        LC020ValidParentheses solution;
        solution = new LC020ValidParentheses();
        assertFalse(solution.isValid("*"));
    }

    @Test
    public void testRB() throws Exception {
        LC020ValidParentheses solution;
        solution = new LC020ValidParentheses();
        assertFalse(solution.isValid("("));
    }

    @Test
    public void testRE() throws Exception {
        LC020ValidParentheses solution;
        solution = new LC020ValidParentheses();
        assertFalse(solution.isValid(")"));
    }

    @Test
    public void testRBRE() throws Exception {
        LC020ValidParentheses solution;
        solution = new LC020ValidParentheses();
        assertTrue(solution.isValid("()"));
    }

    @Test
    public void testRBRESBSECBCE() throws Exception {
        LC020ValidParentheses solution;
        solution = new LC020ValidParentheses();
        assertTrue(solution.isValid("()[]{}"));
    }

    @Test
    public void testRBSBRESE() throws Exception {
        LC020ValidParentheses solution;
        solution = new LC020ValidParentheses();
        assertFalse(solution.isValid("([)]"));
    }

    @Test
    public void testRBCBRECE() throws Exception {
        LC020ValidParentheses solution;
        solution = new LC020ValidParentheses();
        assertFalse(solution.isValid("({)}"));
    }

    @Test
    public void testSBRBSERE() throws Exception {
        LC020ValidParentheses solution;
        solution = new LC020ValidParentheses();
        assertFalse(solution.isValid("[(])"));
    }

    @Test
    public void testCBRBCERE() throws Exception {
        LC020ValidParentheses solution;
        solution = new LC020ValidParentheses();
        assertFalse(solution.isValid("{(})"));
    }

    @Test
    public void testRESECE() throws Exception {
        LC020ValidParentheses solution;
        solution = new LC020ValidParentheses();
        assertFalse(solution.isValid(")"));
        assertFalse(solution.isValid("]"));
        assertFalse(solution.isValid("}"));
    }

    @Test
    public void testNothing() throws Exception {
        LC020ValidParentheses solution;
        solution = new LC020ValidParentheses();
        assertFalse(solution.isValid(null));
        assertTrue(solution.isValid(""));
    }
}
